system_zero/src/permutation_unit.rs

```diff
@@ -2,36 +2,117 @@ use plonky2::field::extension_field::{Extendable, FieldExtension};
 use plonky2::field::packed_field::PackedField;
 use plonky2::hash::hash_types::RichField;
 use plonky2::hash::hashing::SPONGE_WIDTH;
+use plonky2::hash::poseidon::{HALF_N_FULL_ROUNDS, N_PARTIAL_ROUNDS};
 use plonky2::plonk::circuit_builder::CircuitBuilder;
 use starky::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
 use starky::vars::StarkEvaluationTargets;
 use starky::vars::StarkEvaluationVars;
 
-use crate::column_layout::{col_permutation_input, col_permutation_output, NUM_COLUMNS};
+use crate::column_layout::{
+    col_permutation_full_first_after_mds, col_permutation_full_first_mid_sbox,
+    col_permutation_full_second_after_mds, col_permutation_full_second_mid_sbox,
+    col_permutation_input, col_permutation_partial_after_sbox, col_permutation_partial_mid_sbox,
+    NUM_COLUMNS,
+};
 use crate::public_input_layout::NUM_PUBLIC_INPUTS;
 use crate::system_zero::SystemZero;
 
+fn constant_layer<F, FE, P, const D2: usize>(
+    mut state: [P; SPONGE_WIDTH],
+    round: usize,
+) -> [P; SPONGE_WIDTH]
+where
+    F: RichField,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    // One day I might actually vectorize this, but today is not that day.
+    for i in 0..P::WIDTH {
+        let mut unpacked_state = [P::Scalar::default(); SPONGE_WIDTH];
+        for j in 0..SPONGE_WIDTH {
+            unpacked_state[j] = state[j].as_slice()[i];
+        }
+        F::constant_layer_field(&mut unpacked_state, round);
+        for j in 0..SPONGE_WIDTH {
+            state[j].as_slice_mut()[i] = unpacked_state[j];
+        }
+    }
+    state
+}
+
+fn mds_layer<F, FE, P, const D2: usize>(mut state: [P; SPONGE_WIDTH]) -> [P; SPONGE_WIDTH]
+where
+    F: RichField,
+    FE: FieldExtension<D2, BaseField = F>,
+    P: PackedField<Scalar = FE>,
+{
+    for i in 0..P::WIDTH {
+        let mut unpacked_state = [P::Scalar::default(); SPONGE_WIDTH];
+        for j in 0..SPONGE_WIDTH {
+            unpacked_state[j] = state[j].as_slice()[i];
+        }
+        unpacked_state = F::mds_layer_field(&unpacked_state);
+        for j in 0..SPONGE_WIDTH {
+            state[j].as_slice_mut()[i] = unpacked_state[j];
+        }
+    }
+    state
+}
+
 impl<F: RichField + Extendable<D>, const D: usize> SystemZero<F, D> {
-    pub(crate) fn generate_permutation_unit(&self, values: &mut [F; NUM_COLUMNS]) {
+    pub(crate) fn generate_permutation_unit(values: &mut [F; NUM_COLUMNS]) {
```

Comment:
Not that this should be a major cost anyway, but maybe it would be good to short-circuit if the input is zeros. I expect that will be common - maybe 50% or more of cycles won't use the permutation unit.

(I mean only for witness generation; the eval methods should generally receive "random" looking inputs since they come from evaluating witness polys on a different domain.)

Reply:
I'd rather do this optimization if we find it actually speeds things up. I'd have to keep the hash of `[0; 12]` in a constant, which would then have to be kept up to date with Poseidon parameters.